### PR TITLE
Feature/more gst compat

### PR DESCRIPTION
**Pull request by pgab**

### CODE REVIEW

File: src/oracle/connection/stmt.rs [hunk 1]
```diff
@@ -17,13 +17,21 @@ pub struct Statement {
...
 }
 
 const NUM_ELEMENTS: usize = 20;
 
 impl Statement {
     pub fn prepare(raw_connection: &Rc<RawConnection>, sql: &str) -> QueryResult<Self> {
-        let mysql = sql.to_string();
+        let mut mysql = sql.to_string();
+        // TODO: this can go wrong: `UPDATE table SET k='LIMIT';`
+        if mysql.contains("LIMIT") {
```
[the remark this answers is not in the document]
Comment (pgab) [reply src/oracle/connection/stmt.rs]: We need to talk to @weiznich for that. Maybe it was a misunderstanding and we would need to implement rather `first` than `limit`. I was already wondering why the here provided `limitClause` was not taken into consideration.

File: src/oracle/connection/mod.rs [hunk 2]
```diff
@@ -71,7 +110,8 @@ impl Connection for OciConnection {
     where
         T: QueryFragment<Self::Backend> + QueryId,
     {
-        let stmt = try!(self.prepare_query(source));
+        // TODO: FIXME: This doesn't work, enjoy debugging it!
```
[the remark this answers is not in the document]
Comment (pgab) [reply src/oracle/connection/mod.rs]: This always returns 0. Imho everything is right. I remember similar problems on GST side, some years ago.

File: src/oracle/connection/stmt.rs [hunk 3]
```diff
@@ -254,10 +283,28 @@ impl Statement {
                         ffi::OCI_ATTR_CHAR_SIZE,
                         self.connection.env.error_handle,
                     );
-                    Self::check_error(self.connection.env.error_handle, status)?;
+                    Self::check_error_sql(self.connection.env.error_handle, status, &self.mysql, "RETRIEVING LENGTH".to_string())?;
                     //tpe_size += 1;
                     tpe = ffi::SQLT_STR;
                 }
+                ffi::SQLT_ODT
```
[the remark this answers is not in the document]
Comment (pgab) [reply src/oracle/connection/stmt.rs]: https://docs.oracle.com/en/database/oracle/oracle-database/12.2/lnoci/data-types.html#GUID-7DA48B90-07C7-41A7-BC57-D8F358A4EEBE

File: src/oracle/connection/stmt.rs [hunk 4]
```diff
@@ -17,13 +17,20 @@ pub struct Statement {
...
 }
 
 const NUM_ELEMENTS: usize = 20;
 
 impl Statement {
     pub fn prepare(raw_connection: &Rc<RawConnection>, sql: &str) -> QueryResult<Self> {
-        let mysql = sql.to_string();
+        let mut mysql = sql.to_string();
+        // TODO: this can go wrong: `UPDATE table SET k='LIMIT';`
+        if let Some(pos) = mysql.find("LIMIT") {
```
Comment (weiznich) [inline src/oracle/connection/stmt.rs]: Diesel does also support [offset](http://docs.diesel.rs/diesel/query_dsl/trait.QueryDsl.html#method.offset) so this should also be handled here.

@sgrif: Maybe that's something we could adjust in diesel 2.0? A solution would be to pack limit and offset together. So instead of the current `SelectStatement` something like this?
```rust
pub struct SelectStatement<
    From,
    Select = DefaultSelectClause,
    Distinct = NoDistinctClause,
    Where = NoWhereClause,
    Order = NoOrderClause,
    LimitOffset = NoLimitOffsetClause,
    GroupBy = NoGroupByClause,
    Locking = NoLockingClause,
> {
    pub(crate) select: Select,
    pub(crate) from: From,
    pub(crate) distinct: Distinct,
    pub(crate) where_clause: Where,
    pub(crate) order: Order,
    pub(crate) limitOffset: LimitOffset,
    pub(crate) offset: Offset,
    pub(crate) group_by: GroupBy,
    pub(crate) locking: Locking,
}
```

This would make it possible to implement that outside of diesel itself (given the re-rebalance coherence rfc)

File: src/oracle/connection/stmt.rs [hunk 5]
```diff
@@ -68,11 +85,26 @@ impl Statement {
             connection: raw_connection.clone(),
             inner_statement: stmt,
             bind_index: 0,
-            // TODO: this can go wrong: `UPDATE table SET k='select';`
-            is_select: sql.contains("SELECT") || sql.contains("select"),
+            // TODO: this can go wrong: `UPDATE table SET k='select';` OR
+            // ```
+            //            CREATE OR REPLACE FORCE VIEW full_bounding_boxes(id, o_c1, o_c2, o_c3, u_c1, u_c2, u_c3, v_c1, v_c2, v_c3, w_c1, w_c2, w_c3)
```
Comment (weiznich) [inline src/oracle/connection/stmt.rs]: Why is this sql fragment here?

File: src/oracle/connection/stmt.rs [hunk 6]
```diff
@@ -135,8 +180,16 @@ impl Statement {
...
+            Self::check_error_sql(
+                self.connection.env.error_handle,
+                status,
+                &self.mysql,
+                "EXECUTING STMT".to_string(),
+            )?;
         }
+        // TODO: since we have a statement cache, which I couldn't find to turn off,
+        // we need to reset the bind_index once we executed the statement
+        self.bind_index = 0;
```
Comment (weiznich) [inline src/oracle/connection/stmt.rs]: Why is this needed? I do not see where bind_index is set to a not 0 value?

File: src/oracle/connection/transaction.rs [hunk 7]
```diff
@@ -100,10 +100,7 @@ impl TransactionManager<OciConnection> for OCITransactionManager {
             };
             Ok(())
         } else {
-            conn.batch_execute(&format!(
-                "COMMIT diesel_savepoint_{}",
-                transaction_depth - 1
-            ))
+            conn.batch_execute("COMMIT")
```
Comment (weiznich) [inline src/oracle/connection/transaction.rs]: What about using `RELEASE diesel_savepoint_{}` here? This should give us a similar semantic like on postgres?

File: src/oracle/query_builder/limit_clause.rs [hunk 8]
```diff
@@ -0,0 +1,19 @@
+/// TODO: this is currently not used, IDK why diesel still takes the original limit clause
```
Comment (weiznich) [inline src/oracle/query_builder/limit_clause.rs]: Diesel just don't know about this type. I think we could make this (somehow) work.

Short version:
We need to define our own [limit function](https://github.com/diesel-rs/diesel/blob/master/diesel/src/query_dsl/mod.rs#L729) that returns our own type. The best way to do this is to define a new trait and implement that for all T that implement `QueryDsl`. After that we could use this (new) function everywhere where we explicitly call limit. The only place where this won't work is in the implementation of `first` because that is inside of diesel itself.

File: src/oracle/query_builder/subselect.rs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+use diesel::expression::array_comparison::MaybeEmpty;
+use diesel::expression::*;
+use diesel::query_builder::*;
+use diesel::result::QueryResult;
+
+use super::Oracle;
+
+#[derive(Debug, Copy, Clone, QueryId)]
+pub struct Subselect<T, ST> {
```
Comment (weiznich) [inline src/oracle/query_builder/subselect.rs]: Why does we need that?

File: src/oracle/query_builder/limit_clause.rs (hunk 8 above)
Comment (sgrif) [reply src/oracle/query_builder/limit_clause.rs]: Once specialization lands you should also just be able to specialize the original impl. We can already do that today if we're ok with making this crate nightly only.

File: src/oracle/connection/cursor.rs [hunk 10]
```diff
@@ -84,6 +84,7 @@ where
             if let Some(err) =
                 Statement::check_error(self.stmt.connection.env.error_handle, status).err()
             {
+                println!("{:?}", self.stmt.mysql);
```
[the remark this answers is not in the document]
Comment (pgab) [reply src/oracle/connection/cursor.rs]: This came to my mind while using it in our code. At many points it isn't obvious which SQL is behind a query. This easy if you start a new application from scratch. In our case we had a "ready" application with hundreds of requests and also a bunch of tests. Reading the error e.g. `ambiguous column definition` after a test has failed won't help you much. Therefore I added `println` e.g. here. What would be a Rust way for something like `#ifdef DEBUG`?

File: src/oracle/connection/stmt.rs (hunk 5 above)
Comment (pgab) [reply src/oracle/connection/stmt.rs]: Added more explanation.

File: src/oracle/connection/stmt.rs (hunk 6 above)
Comment (pgab) [reply src/oracle/connection/stmt.rs]: This happens when the statement is cached, which occurs if you try to execute the same statement with varying parameters. Since the binding is re-executed the internal counter will continue counting where it stopped before. Setting this to 0 after a statement has been run will start with 1 if new parameters are used with an existing statement.

File: src/oracle/connection/transaction.rs (hunk 7 above)
Comment (pgab) [reply src/oracle/connection/transaction.rs]: I am sorry, what do you mean? Afaik there is no `RELEASE` for Oracle. What do you want to do?

File: src/oracle/query_builder/limit_clause.rs (hunk 8 above)
Comment (pgab) [reply src/oracle/query_builder/limit_clause.rs]: @weiznich I am not sure what you mean. We might discuss that via chat or in person.

File: src/oracle/query_builder/subselect.rs (hunk 9 above)
Comment (pgab) [reply src/oracle/query_builder/subselect.rs]: We could define our own expression (`exists`) but then would probably need to implement everything here instead of using diesel (this yields some compiler errors about unfulfilled traits) (c.f. https://github.com/GiGainfosystems/diesel-oci/blob/feature/more_gst_compat/src/oracle/query_builder/exists.rs#L48)

File: src/oracle/connection/transaction.rs (hunk 7 above)
Comment (weiznich) [reply src/oracle/connection/transaction.rs]: [This](https://docs.oracle.com/cd/E17952_01/mysql-5.1-en/savepoint.html) mentions a `RELEASE savepoint` syntax.

Comment (pgab) [reply src/oracle/connection/transaction.rs]: Thats mysql ;)
c.f.
https://docs.oracle.com/database/121/SQLRF/statements_9.htm#SQLRF016
https://docs.oracle.com/database/121/CNCPT/transact.htm#CNCPT038
https://docs.oracle.com/database/121/CNCPT/sqllangu.htm#CNCPT1733

Comment (pgab) [reply src/oracle/connection/transaction.rs]: I'll give your idea a thought. We might need to define/discuss some edge cases here and then test those. Since as of now we use for the "base" transaction OCI and for all other nested transactions SQL. But your idea sounds doable.

File: src/oracle/connection/mod.rs [hunk 11]
```diff
@@ -25,6 +27,33 @@ pub struct OciConnection {
     raw: Rc<raw::RawConnection>,
     transaction_manager: OCITransactionManager,
     statement_cache: StatementCache<Oracle, Statement>,
+    has_open_test_transaction: Cell<bool>,
+}
+
+impl MigrationConnection for OciConnection {
+    #[cfg(ka)]
```
Comment (weiznich) [inline src/oracle/connection/mod.rs]: That should not be here.

File: src/oracle/connection/mod.rs [hunk 12]
```diff
@@ -71,7 +109,8 @@ impl Connection for OciConnection {
     where
         T: QueryFragment<Self::Backend> + QueryId,
     {
-        let stmt = try!(self.prepare_query(source));
+        // TODO: FIXME: this always returns 0 whereas the code looks proper
```
Comment (weiznich) [inline src/oracle/connection/mod.rs]: There should be an open issue for this.

File: src/oracle/connection/stmt.rs [hunk 13]
```diff
@@ -17,13 +17,20 @@ pub struct Statement {
     buffers: Vec<Box<[u8]>>,
     sizes: Vec<i32>,
     indicators: Vec<Box<ffi::OCIInd>>,
+    pub mysql: String,
```
Comment (weiznich) [inline src/oracle/connection/stmt.rs]: I do not think this needs to be public, prefer `pub(crate)` if it's used internally.

File: src/oracle/connection/stmt.rs (hunk 4 above)
Comment (weiznich) [inline src/oracle/connection/stmt.rs]: We should have an open issue for this, so we have at least some sort of documentation what may be broken and what should be fixed as soon as certain language features arrive.

File: src/oracle/connection/stmt.rs [hunk 14]
```diff
@@ -122,7 +157,20 @@ impl Statement {
         }
     }
 
-    pub fn run(&self) -> QueryResult<()> {
+    pub fn check_error_sql(
+        error_handle: *mut ffi::OCIError,
+        status: i32,
+        sql: &String,
+        action: String,
```
Comment (weiznich) [inline src/oracle/connection/stmt.rs]: Do we really need a `String` here, or could we just use a `&str` to skip some allocations?

(This would allow removing all those `.to_string()` calls while calling this function)

File: src/oracle/connection/stmt.rs (hunk 6 above)
Comment (weiznich) [reply src/oracle/connection/stmt.rs]: The comment should say something like this:
"The bind index is required to start by zero. If a statement is executed more than once we need to reset the index here"

(The comment should not mention something about the statement cache, because we want this cache :wink:)

File: src/oracle/query_builder/alias.rs [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    }
+}
+
+#[derive(Debug, Clone, QueryId)]
+pub struct As<T> {
+    query: T,
+    alias: String,
+}
+
+use diesel::expression::Expression;
```
Comment (weiznich) [inline src/oracle/query_builder/alias.rs]: Move all those imports to the top of the file.

File: src/oracle/query_builder/exists.rs [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    // this doesn't work for oracle (is always 0)
+    //let cnt = conn.execute_returning_count(&query)?;
+    //if cnt > 0 {
+    //    return Ok(true);
+    //} else {
+    //    return Ok(false);
+    //}
+    // TODO: I am not happy with that, but currently diesel has not other options
+    // the code above would be better, need to check why it doesn't work
+    let v: Vec<U> = conn.query_by_index(query)?;
```
Comment (weiznich) [inline src/oracle/query_builder/exists.rs]: Didn't we talked about using `select count(*)` here?

File: src/oracle/query_dsl/save_changes_dsl.rs [hunk 17]
```diff
@@ -7,21 +7,21 @@ use diesel::query_dsl::methods::{ExecuteDsl, FindDsl};
 use diesel::query_dsl::{LoadQuery, RunQueryDsl};
 use diesel::result::QueryResult;
 
-use super::super::OciConnection;
+use super::super::connection::OciConnection;
```
Comment (weiznich) [inline src/oracle/query_dsl/save_changes_dsl.rs]: Maybe a absolute import would be better?

File: src/test/mod.rs [hunk 18]
```diff
@@ -1,12 +1,19 @@
...
 use std::env;
 
+fn init_testing() -> OciConnection {
+    use super::logger::init;
+    init();
+    //let database_url = database_url_from_env("OCI_DATABASE_URL");
+    OciConnection::establish(&DB_URL).expect("No connection, no test!")
+}
+
 #[cfg(ka)]
```
Comment (weiznich) [inline src/test/mod.rs]: Remove this and the impl below

File: src/test/mod.rs [hunk 19]
```diff
@@ -55,7 +62,6 @@ fn database_url_from_env(backend_specific_env_var: &str) -> String {
     }
 }
 
-
 const DB_URL: &str = "oci://\"diesel\"/diesel@//192.168.2.81:1521/orcl";
```
Comment (weiznich) [inline src/test/mod.rs]: We should not hardcode our internal ip here :wink:

File: src/test/mod.rs [hunk 20]
```diff
@@ -217,10 +220,143 @@ fn transaction_rollback() {
...
+
+        let out_inner = conn.transaction::<_, Error, _>(|| {
+            let sql_inner = format!("INSERT INTO test ({}) VALUES ({})", "TST_CHR", TEST_VARCHAR);
+            let _ret_inner = conn.execute(&*sql_inner)?;
+            let ret_inner =
+                self::test::dsl::test.load::<(Option<i64>, Option<String>, Option<i64>)>(&conn)?;
+            assert_eq!(ret_inner.len(), 2);
+            Ok(())
+        });
+        assert_result!(out_inner);
```
Comment (weiznich) [inline src/test/mod.rs]: There should be a second inner transaction that is rolled back.

File: src/test/mod.rs [hunk 21]
```diff
@@ -324,6 +460,66 @@ enum Way {
     Native,
 }
 
+extern crate chrono;
+use self::chrono::NaiveDateTime;
```
Comment (weiznich) [inline src/test/mod.rs]: Move this two lines to the top of the file.

File: src/test/mod.rs [hunk 22]
```diff
@@ -508,6 +734,927 @@ fn gst_compat() {
...
+            Option<Vec<u8>>,
+        ),
+        Error,
+    > = gst_types
+        .select((big, small, normal, d, r, v, byte))
+        .first(&conn);
+    assert_result!(ret);
+}
+
+use diesel::deserialize::{self, FromSql};
```
Comment (weiznich) [inline src/test/mod.rs]: Imports should be on the top of the current scope. (This means on the top of this file in this case)

File: src/test/mod.rs [hunk 23]
```diff
@@ -508,6 +734,927 @@ fn gst_compat() {
...
+
+table! {
+    /// all tables
+    all_tables (owner, table_name) {
+        /// all tables
+        owner -> Text,
+        /// all tables
+        table_name -> Text,
+//        /// all tables
+//        tablespace_name -> Nullable<Text>,
```
Comment (weiznich) [inline src/test/mod.rs]: Remove all those commented out lines/columns.

File: src/test/mod.rs (hunk 19 above)
Comment (pgab) [reply src/test/mod.rs]: In general I do agree with you, but it's already out, second this is a VM/docker which just exists for that one task. So there is no harm anyway. Removed it, please remember to set the environment variable `OCI_DATABASE_URL` from now on.

File: src/test/mod.rs (hunk 20 above)
Comment (pgab) [reply src/test/mod.rs]: We insert something in the first transaction `a`. Then we open another transaction `b` and insert another value. After that we commit `b`. So `a` sees the change. Then we rollback `a` with an arbitrary error. The initial connection doesn't see anything of neither `a` nor `b`. Why would we need another inner transaction `c`? Should that go in `a` or in `b`?

Comment (weiznich) [reply src/test/mod.rs]: We need a second inner transaction because we know about the inner details of how nested transactions are implemented. The outer level uses the oci level transaction, whereas the inner transaction uses savepoints. So we need to check that both rollback variants are working.

File: src/test/mod.rs (hunk 19 above)
Comment (weiznich) [reply src/test/mod.rs]: It's not about that this url is secret, it is just about that other people could test this crate without setting up a database at a specific ip :wink:

File: src/test/mod.rs (hunk 20 above)
Comment (pgab) [reply src/test/mod.rs]: I see what you mean. I think we should wrap this in new tests.
